File: ci/discovery.py
#!/usr/bin/env nix
#! nix shell --inputs-from .# nixpkgs#python3 --command python3
"""Build the GitHub Actions update matrix for packages and flake inputs."""

from __future__ import annotations

import json
import os
from dataclasses import asdict, dataclass
from pathlib import Path

from lib import NATIVE_RUNNERS, run, write_output

PACKAGE_SYSTEMS = tuple(NATIVE_RUNNERS)
MANUAL_UPDATE_MARKER = "no-auto-update"
PACKAGE_EXPRESSION = r"""
let
  config = builtins.fromJSON (builtins.getEnv "DISCOVERY_CONFIG");
  flake = builtins.getFlake (toString ./.);
  pkgs = flake.packages.${config.system} or { };
  isHidden = pkg: (builtins.tryEval (pkg.passthru.hideFromDocs or false)).value or false;
  shouldDiscover = pkg: !(isHidden pkg) && pkg ? version;
in
  builtins.mapAttrs (_: pkg: if shouldDiscover pkg then pkg.version else null) pkgs
"""
GROUP_SPECS = (
    ("linux-x86", "x86_64-linux", "package"),
    ("linux-arm", "aarch64-linux", "package"),
    ("darwin-arm", "aarch64-darwin", "package"),
    ("linux-flake-input", "x86_64-linux", "flake-input"),
)


@dataclass(frozen=True, slots=True)
class Target:
    current_version: str
    name: str
    system: str
    type: str


@dataclass(frozen=True, slots=True)
class MatrixItem:
    group: str
    runner: str
    system: str
    targets: list[dict[str, str]]
    type: str


def split_filter(value: str) -> list[str] | None:
    items = value.split()
    return items or None


def package_versions(system: str) -> dict[str, str]:
    result = run(
        ["nix", "eval", "--json", "--impure", "--expr", PACKAGE_EXPRESSION],
        capture=True,
        check=False,
        env={"DISCOVERY_CONFIG": json.dumps({"system": system})},
    )
    if result.returncode != 0:
        raise RuntimeError(
            f"Failed to discover packages for {system}:\n{result.stderr}"
        )

    payload: object = json.loads(result.stdout)
    if not isinstance(payload, dict):
        raise RuntimeError(
            f"Package discovery for {system} returned a non-object JSON value"
        )
    return {
        name: version
        for name, version in payload.items()
        if isinstance(name, str) and isinstance(version, str)
    }


def automatic_updates_disabled(name: str) -> bool:
    return (Path("packages") / name / MANUAL_UPDATE_MARKER).is_file()


def discover_packages(package_filter: list[str] | None) -> list[Target]:
    discovered: dict[str, Target] = {}
    disabled: set[str] = set()
    for system in PACKAGE_SYSTEMS:
        for name, version in package_versions(system).items():
            if automatic_updates_disabled(name):
                disabled.add(name)
                continue
            if package_filter is None or name in package_filter:
                discovered.setdefault(
                    name,
                    Target(
                        current_version=version,
                        name=name,
                        system=system,
                        type="package",
                    ),
                )

    for name in package_filter or []:
        if name in disabled:
            print(f"::warning::Package {name} has automatic updates disabled")
        elif name not in discovered:
            print(f"::warning::Package {name} was not found or has no version")
    return sorted(discovered.values(), key=lambda item: item.name)


def discover_flake_inputs(input_filter: list[str] | None) -> list[Target]:
    lock_path = Path("flake.lock")
    if not lock_path.exists():
        return []

    lock = json.loads(lock_path.read_text())
    if not isinstance(lock, dict):
        raise RuntimeError("flake.lock is not a JSON object")
    nodes = lock.get("nodes")
    if not isinstance(nodes, dict):
        raise RuntimeError("flake.lock has no nodes")
    root = nodes.get("root")
    root_inputs = root.get("inputs") if isinstance(root, dict) else None
    if not isinstance(root_inputs, dict):
        raise RuntimeError("flake.lock has no root inputs")

    names = input_filter or sorted(root_inputs)
    items: list[Target] = []
    for name in names:
        node = nodes.get(name)
        locked = node.get("locked") if isinstance(node, dict) else None
        revision = (
            locked.get("rev") or locked.get("lastModified")
            if isinstance(locked, dict)
            else None
        )
        items.append(
            Target(
                current_version=str(revision)[:8]
                if revision is not None
                else "unknown",
                name=name,
                system="x86_64-linux",
                type="flake-input",
            )
        )
    return items


def encode_targets(items: list[Target]) -> list[dict[str, str]]:
    return [
        {"current_version": item.current_version, "name": item.name} for item in items
    ]


def build_matrix(
    packages: list[Target], flake_inputs: list[Target]
) -> dict[str, list[dict[str, object]]]:
    include: list[dict[str, object]] = []
    for group, system, update_type in GROUP_SPECS:
        items = [
            item
            for item in (packages if update_type == "package" else flake_inputs)
            if item.system == system and item.type == update_type
        ]
        if not items:
            continue
        include.append(
            asdict(
                MatrixItem(
                    group=group,
                    runner=NATIVE_RUNNERS[system],
                    system=system,
                    targets=encode_targets(items),
                    type=update_type,
                )
            )
        )
    return {"include": include}


def main() -> None:
    package_filter = split_filter(os.environ.get("PACKAGES", ""))
    input_filter = split_filter(os.environ.get("INPUTS", ""))
    matrix = build_matrix(
        discover_packages(package_filter),
        discover_flake_inputs(input_filter),
    )
    print(json.dumps(matrix, indent=2))
    write_output("matrix", json.dumps(matrix, separators=(",", ":")))
    write_output("has-updates", str(bool(matrix["include"])).lower())


if __name__ == "__main__":
    main()

File: ci/discovery.ts
import { existsSync } from "node:fs";
import process from "node:process";
import type { SystemConfig } from "./lib.ts";
import {
	compactJson,
	isRecord,
	parseJson,
	prettyJson,
	readJsonFile,
	readSystems,
	run,
	splitFilter,
	writeOutput,
} from "./lib.ts";

const MANUAL_UPDATE_MARKER = "no-auto-update";
const PACKAGE_EXPRESSION = `
let
  config = builtins.fromJSON (builtins.getEnv "DISCOVERY_CONFIG");
  flake = builtins.getFlake (toString ./.);
  pkgs = flake.packages.\${config.system} or { };
  isHidden = pkg: (builtins.tryEval (pkg.passthru.hideFromDocs or false)).value or false;
  shouldDiscover = pkg: !(isHidden pkg) && pkg ? version;
in
  builtins.mapAttrs (_: pkg: if shouldDiscover pkg then pkg.version else null) pkgs
`;

type UpdateType = "flake-input" | "package";

export type Target = Readonly<{
	currentVersion: string;
	name: string;
	system: string;
	type: UpdateType;
}>;

export type MatrixItem = Readonly<{
	artifact: string;
	group: string;
	requires_internet_archive: boolean;
	runner: string;
	system: string;
	target: Readonly<{ current_version: string; name: string }>;
	type: UpdateType;
}>;

export type UpdateMatrix = Readonly<{
	include: readonly MatrixItem[];
}>;

export type BatchMatrixItem = Readonly<{
	artifact: string;
	group: string;
	runner: string;
	targets: readonly MatrixItem[];
}>;

export type BatchMatrix = Readonly<{
	include: readonly BatchMatrixItem[];
}>;

function packageDirectory(name: string): string {
	return `packages/${name}`;
}

function automaticUpdatesDisabled(name: string): boolean {
	return existsSync(`${packageDirectory(name)}/${MANUAL_UPDATE_MARKER}`);
}

export function parsePackageVersions(
	value: unknown,
	system: string,
): ReadonlyMap<string, string> {
	if (!isRecord(value)) {
		throw new Error(`Package discovery for ${system} returned a non-object`);
	}
	const versions = new Map<string, string>();
	for (const [name, version] of Object.entries(value)) {
		if (typeof version === "string") {
			versions.set(name, version);
		}
	}
	return versions;
}

async function packageVersions(
	system: string,
): Promise<ReadonlyMap<string, string>> {
	const result = await run(
		["nix", "eval", "--json", "--impure", "--expr", PACKAGE_EXPRESSION],
		{
			capture: true,
			check: false,
			env: { DISCOVERY_CONFIG: JSON.stringify({ system }) },
		},
	);
	if (!result.success) {
		throw new Error(
			`Failed to discover packages for ${system}:\n${result.stderr}`,
		);
	}
	return parsePackageVersions(
		parseJson(result.stdout, `package discovery for ${system}`),
		system,
	);
}

async function discoverPackages(
	systems: readonly SystemConfig[],
	packageFilter: readonly string[] | undefined,
): Promise<readonly Target[]> {
	const discovered = new Map<string, Target>();
	const disabled = new Set<string>();

	for (const { system } of systems) {
		for (const [name, version] of await packageVersions(system)) {
			if (automaticUpdatesDisabled(name)) {
				disabled.add(name);
				continue;
			}
			if (
				(packageFilter === undefined || packageFilter.includes(name)) &&
				!discovered.has(name)
			) {
				discovered.set(name, {
					currentVersion: version,
					name,
					system,
					type: "package",
				});
			}
		}
	}

	for (const name of packageFilter ?? []) {
		if (disabled.has(name)) {
			console.log(`::warning::Package ${name} has automatic updates disabled`);
		} else if (!discovered.has(name)) {
			console.log(`::warning::Package ${name} was not found or has no version`);
		}
	}
	return [...discovered.values()].sort((left, right) =>
		left.name.localeCompare(right.name),
	);
}

function lockedRevision(
	nodes: Record<string, unknown>,
	nodeName: string,
): string {
	const node = nodes[nodeName];
	if (!isRecord(node) || !isRecord(node.locked)) {
		return "unknown";
	}
	const revision = node.locked.rev ?? node.locked.lastModified;
	return revision === undefined ? "unknown" : String(revision).slice(0, 8);
}

export function parseFlakeInputs(
	lock: unknown,
	inputFilter: readonly string[] | undefined,
): readonly Target[] {
	if (!isRecord(lock)) {
		throw new Error("flake.lock must be an object");
	}
	const nodes = lock.nodes;
	if (!isRecord(nodes)) {
		throw new Error("flake.lock has no nodes");
	}
	const root = nodes.root;
	if (!isRecord(root)) {
		throw new Error("flake.lock has no root node");
	}
	const inputs = root.inputs;
	if (!isRecord(inputs)) {
		throw new Error("flake.lock has no root inputs");
	}

	const names = inputFilter ?? Object.keys(inputs).sort();
	return names.map((name) => {
		const reference = inputs[name];
		const nodeName = typeof reference === "string" ? reference : name;
		return {
			currentVersion: lockedRevision(nodes, nodeName),
			name,
			system: "x86_64-linux",
			type: "flake-input",
		};
	});
}

function matrixTarget(target: Target): MatrixItem["target"] {
	return {
		current_version: target.currentVersion,
		name: target.name,
	};
}

function artifactName(type: UpdateType, name: string): string {
	return `update-${type}-${name.replaceAll(/[^A-Za-z0-9_.-]/g, "-")}`;
}

export function buildMatrix(
	systems: readonly SystemConfig[],
	packages: readonly Target[],
	flakeInputs: readonly Target[],
): UpdateMatrix {
	const include: MatrixItem[] = [];
	for (const { runner, system } of systems) {
		for (const target of packages.filter((item) => item.system === system)) {
			include.push({
				artifact: artifactName("package", target.name),
				group: `package-${target.name}`,
				requires_internet_archive: target.name === "termius",
				runner,
				system,
				target: matrixTarget(target),
				type: "package",
			});
		}
	}

	const runner = systems.find(
		({ system }) => system === "x86_64-linux",
	)?.runner;
	if (!runner && flakeInputs.length > 0) {
		throw new Error(
			"ci/systems.json must define x86_64-linux for flake input updates",
		);
	}
	for (const target of flakeInputs) {
		include.push({
			artifact: artifactName("flake-input", target.name),
			group: `flake-input-${target.name}`,
			requires_internet_archive: false,
			runner: runner ?? "",
			system: "x86_64-linux",
			target: matrixTarget(target),
			type: "flake-input",
		});
	}
	return { include };
}

export function buildBatchMatrix(matrix: UpdateMatrix): BatchMatrix {
	const batches = new Map<string, BatchMatrixItem>();
	for (const item of matrix.include) {
		const key = `${item.runner}:${item.system}`;
		const existing = batches.get(key);
		if (existing) {
			batches.set(key, { ...existing, targets: [...existing.targets, item] });
		} else {
			batches.set(key, {
				artifact: `update-batch-${item.system}`,
				group: item.system,
				runner: item.runner,
				targets: [item],
			});
		}
	}
	return {
		include: [...batches.values()].sort((left, right) =>
			left.group.localeCompare(right.group),
		),
	};
}

export async function discoverUpdates(): Promise<void> {
	const systems = readSystems();
	const packageFilter = splitFilter(process.env.PACKAGES);
	const inputFilter = splitFilter(process.env.INPUTS);
	const matrix = buildMatrix(
		systems,
		await discoverPackages(systems, packageFilter),
		parseFlakeInputs(readJsonFile("flake.lock"), inputFilter),
	);
	const batchMatrix = buildBatchMatrix(matrix);
	console.log(prettyJson(batchMatrix).trimEnd());
	writeOutput("batch-matrix", compactJson(batchMatrix));
	writeOutput("matrix", compactJson(matrix));
	writeOutput("has-updates", String(matrix.include.length > 0));
}
